### ISSUE TITLE
Can't reconnect if the server was restarted

### ISSUE THREAD

**Issue by Imaskar** [nats-io/java-nats#177]
Situation:
Server v1.3, with authentication, client v2.1.0.
Initialization code:
```
Options.Builder builder = new Options.Builder()
        .server(<...>)
        .userInfo(<...>,<...>)
        .maxReconnects(-1)
        .pingInterval(Duration.ofMillis(PING_INTERVAL_MS));
Options opts = builder.build();
Nats nats = Nats.connect(opts);
```
Old behavior (Nats v1.0.4, Jnats v1.0.0):
Server started->client connects->client pushes messages->server restarted->client tries to reconnect->ok->client pushes messages.

New behavior:
Server started->client connects->client pushes messages->server restarted->client tries to reconnect->gets error->can't push.

Server log:

>  [27118] 2018/09/12 12:33:30.236462 [DBG] 127.0.0.1:47154 - cid:10 - Client connection created
> [27118] 2018/09/12 12:33:30.253822 [ERR] 127.0.0.1:47154 - cid:10 - Authorization Error - User ""
> [27118] 2018/09/12 12:33:30.253869 [TRC] 127.0.0.1:47154 - cid:10 - <<- [-ERR Authorization Violation]
> [27118] 2018/09/12 12:33:30.253895 [DBG] 127.0.0.1:47154 - cid:10 - Client connection closed

So, it seems that client relies on authentication cache and doesn't send credentials again.

**Comment by sasbury** [comment]
I will take a look, perhaps something with an earlier change to allow credentials on multiple URLs.

**Comment by sasbury** [comment]
I tried to write a test to duplicate this, can you take a look and either run this or let me know how it differs? The only thing I can think of is hte ping interval, which shouldn't effect auth, but maybe it is somehow.


    @Test
    public void testUserPassOnReconnect() throws Exception {
        TestHandler handler = new TestHandler();
        int port = NatsTestServer.nextPort();
        Connection nc = null;
        Subscription sub = null;
        String[] customArgs = {"--user","stephen","--pass","password"};

        try (NatsTestServer ts = new NatsTestServer(customArgs, port, false)) {
            // See config file for user/pass
            Options options = new Options.Builder().
                        server(ts.getURI()).
                        maxReconnects(-1).
                        userInfo("stephen", "password").
                        connectionListener(handler).
                        build();
            nc = Nats.connect(options);
            assertTrue("Connected Status", Connection.Status.CONNECTED == nc.getStatus());

            sub = nc.subscribe("test");
            nc.publish("test", null);
            nc.flush(Duration.ofSeconds(5));
            Message msg = sub.nextMessage(Duration.ofSeconds(5));
            assertNotNull(msg);
            handler.prepForStatusChange(Events.DISCONNECTED);
        }

        try {
            nc.flush(Duration.ofSeconds(1));
        } catch (Exception exp) {
        }

        handler.waitForStatusChange(5, TimeUnit.SECONDS);
        assertTrue("Reconnecting status", Connection.Status.RECONNECTING == nc.getStatus() ||
                                            Connection.Status.DISCONNECTED == nc.getStatus());
        handler.prepForStatusChange(Events.RESUBSCRIBED);


        try (NatsTestServer ts = new NatsTestServer(customArgs, port, false)) {
            handler.waitForStatusChange(5, TimeUnit.SECONDS);
            assertTrue("Connected Status", Connection.Status.CONNECTED == nc.getStatus());

            nc.publish("test", null);
            nc.flush(Duration.ofSeconds(5));
            Message msg = sub.nextMessage(Duration.ofSeconds(5));
            assertNotNull(msg);

            nc.close();
            assertTrue("Closed Status", Connection.Status.CLOSED == nc.getStatus());
        }
    }

If this works for you then i will dig into the ping possibility.

**Comment by Imaskar** [comment]
/* offtopic
I haven't worked with jnats sources before, and I have to note that spinning them up requires some additional steps:

- comment out signing
- comment out publishing
- export gnatsd path to magic environment variable  I found in sources (GNATSD_PATH)
*/

Now to the point: this test works both for v.1.0.4 and v1.3 servers. Both with ping interval (I also added 2*interval wait before reconnection) and without. I don't get it, gonna recheck my tests.

**Comment by sasbury** [comment]
you can run the tests with ./gradlew test, i don't think you need to comment anything (but if you do please tell me I am wrong, i didn't intend that) The gnatsd is magic, I think I need to doc that more.

Let me know if you can reproduce, if not maybe we can close this for now. I will check in the tests though.

**Comment by Imaskar** [comment]
I've finally been able to identify what's going on. **TL/DR: after reconnecting client tries to flush buffer before authenticating.**

Client pushes messages->server stopped->client can't send, so keeps a buffer->server started->[client connects->tries to flush a portion of buffer->error->disconnects->waits some timeout->]*->connects->pushes messages.

If the client pushes messages rapidly, buffer can grow large and that repeated part can take a long time. More so, I don't know where those messages sent after starting the server go. If they go in the same buffer, it can grow faster than failed uploads discard portions (I think this was our case).

It looks like this in log:

```
[2031] 2018/09/18 12:54:33.219869 [TRC] 127.0.0.1:54032 - cid:4 - ->> [CONNECT {"lang":"java","version":"2.0.0","protocol":1,"verbose":false,"pedantic":false,"tls_required":false,"echo":true,"user":"...","pass":"..."}]
[2031] 2018/09/18 12:54:33.219921 [TRC] 127.0.0.1:54032 - cid:4 - ->> [PING]
[2031] 2018/09/18 12:54:33.219929 [TRC] 127.0.0.1:54032 - cid:4 - <<- [PONG]
[2031] 2018/09/18 12:54:33.223722 [TRC] 127.0.0.1:54032 - cid:4 - ->> [SUB * 1]
[2031] 2018/09/18 12:54:33.223764 [TRC] 127.0.0.1:54032 - cid:4 - ->> [PUB msg1 0]
[2031] 2018/09/18 12:54:33.223771 [TRC] 127.0.0.1:54032 - cid:4 - ->> MSG_PAYLOAD: []
[2031] 2018/09/18 12:54:33.223782 [TRC] 127.0.0.1:54032 - cid:4 - <<- [MSG msg1 1 0]
[2031] 2018/09/18 12:54:33.238167 [TRC] 127.0.0.1:54032 - cid:4 - ->> [PING]
[2031] 2018/09/18 12:54:33.238195 [TRC] 127.0.0.1:54032 - cid:4 - <<- [PONG]
[2031] 2018/09/18 12:54:33.239082 [TRC] 127.0.0.1:54032 - cid:4 - ->> [PING]
[2031] 2018/09/18 12:54:33.239101 [TRC] 127.0.0.1:54032 - cid:4 - <<- [PONG]
[2031] 2018/09/18 12:54:34.221444 [TRC] 127.0.0.1:54032 - cid:4 - ->> [PING]
[2031] 2018/09/18 12:54:34.221495 [TRC] 127.0.0.1:54032 - cid:4 - <<- [PONG]
[2143] 2018/09/18 12:54:36.031669 [INF] Starting nats-server version 1.3.0
[2143] 2018/09/18 12:54:36.034675 [DBG] Go build version go1.11
[2143] 2018/09/18 12:54:36.034854 [INF] Git commit [not set]
[2143] 2018/09/18 12:54:36.035541 [INF] Starting http monitor on 0.0.0.0:8222
[2143] 2018/09/18 12:54:36.035746 [INF] Listening for client connections on 0.0.0.0:8240
[2143] 2018/09/18 12:54:36.035856 [DBG] Server id is qhwiL7MIoK2q1MySrHqvTy
[2143] 2018/09/18 12:54:36.035894 [INF] Server is ready
[2143] 2018/09/18 12:54:36.908542 [DBG] 127.0.0.1:54040 - cid:1 - Client connection created
[2143] 2018/09/18 12:54:36.917824 [ERR] 127.0.0.1:54040 - cid:1 - Authorization Error - User ""
[2143] 2018/09/18 12:54:36.917961 [TRC] 127.0.0.1:54040 - cid:1 - <<- [-ERR Authorization Violation]
[2143] 2018/09/18 12:54:36.918063 [DBG] 127.0.0.1:54040 - cid:1 - Client connection closed
[2143] 2018/09/18 12:54:40.924373 [DBG] 127.0.0.1:54042 - cid:2 - Client connection created
[2143] 2018/09/18 12:54:40.929376 [ERR] 127.0.0.1:54042 - cid:2 - Authorization Error - User ""
[2143] 2018/09/18 12:54:40.929477 [TRC] 127.0.0.1:54042 - cid:2 - <<- [-ERR Authorization Violation]
[2143] 2018/09/18 12:54:40.929564 [DBG] 127.0.0.1:54042 - cid:2 - Client connection closed
[2143] 2018/09/18 12:54:44.934192 [DBG] 127.0.0.1:54044 - cid:3 - Client connection created
[2143] 2018/09/18 12:54:44.941690 [ERR] 127.0.0.1:54044 - cid:3 - Authorization Error - User ""
[2143] 2018/09/18 12:54:44.941790 [TRC] 127.0.0.1:54044 - cid:3 - <<- [-ERR Authorization Violation]
[2143] 2018/09/18 12:54:44.941900 [DBG] 127.0.0.1:54044 - cid:3 - Client connection closed
[2143] 2018/09/18 12:54:48.944582 [DBG] 127.0.0.1:54048 - cid:4 - Client connection created
[2143] 2018/09/18 12:54:48.947632 [TRC] 127.0.0.1:54048 - cid:4 - ->> [CONNECT {"lang":"java","version":"2.0.0","protocol":1,"verbose":false,"pedantic":false,"tls_required":false,"echo":true,"user":"...","pass":"..."}]
[2143] 2018/09/18 12:54:48.947833 [TRC] 127.0.0.1:54048 - cid:4 - ->> [CONNECT {"lang":"java","version":"2.0.0","protocol":1,"verbose":false,"pedantic":false,"tls_required":false,"echo":true,"user":"...","pass":"..."}]
[2143] 2018/09/18 12:54:48.947905 [TRC] 127.0.0.1:54048 - cid:4 - ->> [CONNECT {"lang":"java","version":"2.0.0","protocol":1,"verbose":false,"pedantic":false,"tls_required":false,"echo":true,"user":"...","pass":"..."}]
[2143] 2018/09/18 12:54:48.947984 [TRC] 127.0.0.1:54048 - cid:4 - ->> [CONNECT {"lang":"java","version":"2.0.0","protocol":1,"verbose":false,"pedantic":false,"tls_required":false,"echo":true,"user":"...","pass":"..."}]
[2143] 2018/09/18 12:54:48.948036 [TRC] 127.0.0.1:54048 - cid:4 - ->> [PING]
[2143] 2018/09/18 12:54:48.948050 [TRC] 127.0.0.1:54048 - cid:4 - <<- [PONG]
[2143] 2018/09/18 12:54:48.950298 [TRC] 127.0.0.1:54048 - cid:4 - ->> [SUB * 1]
[2143] 2018/09/18 12:54:48.993601 [TRC] 127.0.0.1:54048 - cid:4 - ->> [PING]
[2143] 2018/09/18 12:54:48.993655 [TRC] 127.0.0.1:54048 - cid:4 - <<- [PONG]
[2143] 2018/09/18 12:54:48.994525 [TRC] 127.0.0.1:54048 - cid:4 - ->> [PUB msg5 0]
[2143] 2018/09/18 12:54:48.994563 [TRC] 127.0.0.1:54048 - cid:4 - ->> MSG_PAYLOAD: []
[2143] 2018/09/18 12:54:48.994617 [TRC] 127.0.0.1:54048 - cid:4 - <<- [MSG msg5 1 0]
```
I don't know why it logged connection 4 times, I haven't seen this before. The problem is failed attempts to resend buffer. In this example there were 6k messages buffered before reconnecting.

For now I haven't come up with a test that would highlight it, because a client resubscribes much later than reconnecting. But it is clearly visible in log.

**Comment by Imaskar** [comment]
Something like this:

```
@Test
  public void testBufferingWhileDisconnected() throws Exception {
    TestHandler handlerConsumer = new TestHandler();
    TestHandler handlerProducer = new TestHandler();
    int port = NatsTestServer.nextPort();
    Connection consumer = null, producer = null;
    Subscription sub = null;
    String[] customArgs = {"--user","stephen","--pass","password"};

    try (NatsTestServer ts = new NatsTestServer(customArgs, port, true)) {
      Options.Builder options = new Options.Builder().
          server(ts.getURI()).
          maxReconnects(-1).
          userInfo("stephen", "password").
          connectionListener(handlerConsumer);
      consumer = Nats.connect(options.build());
      options.connectionListener(handlerProducer);
      producer = Nats.connect(options.build());
      assertTrue("Connected Status", Connection.Status.CONNECTED == producer.getStatus());
      assertTrue("Connected Status", Connection.Status.CONNECTED == consumer.getStatus());

      sub = consumer.subscribe("before");
      consumer.flush(Duration.ofSeconds(1));
      producer.publish("before", null);
      producer.flush(Duration.ofSeconds(5));
      Message msg = sub.nextMessage(Duration.ofSeconds(5));
      assertNotNull(msg);
      handlerProducer.prepForStatusChange(ConnectionListener.Events.DISCONNECTED);
    }
    try {
      producer.flush(Duration.ofSeconds(1));
    } catch (Exception exp) {
    }

    handlerProducer.waitForStatusChange(5, TimeUnit.SECONDS);
    Connection.Status status = producer.getStatus();
    assertTrue("Reconnecting status", Connection.Status.RECONNECTING == status ||
        Connection.Status.DISCONNECTED == status);
    for (int i = 0; i < 6000; i++) {
      producer.publish("before", null);
    }
    handlerProducer.prepForStatusChange(ConnectionListener.Events.RESUBSCRIBED);
    handlerConsumer.prepForStatusChange(ConnectionListener.Events.RESUBSCRIBED);

    try (NatsTestServer ts = new NatsTestServer(customArgs, port, true)) {
      handlerProducer.waitForStatusChange(5, TimeUnit.SECONDS);
      handlerConsumer.waitForStatusChange(5, TimeUnit.SECONDS);
      assertTrue("Connected Status", Connection.Status.CONNECTED == producer.getStatus());
      assertTrue("Connected Status", Connection.Status.CONNECTED == consumer.getStatus());

      producer.publish("after", null);
      producer.flush(Duration.ofSeconds(5));
      Message msg = sub.nextMessage(Duration.ofSeconds(5));
      assertNotNull(msg);

      producer.close();
      consumer.close();
      assertTrue("Closed Status", Connection.Status.CLOSED == producer.getStatus());
      assertTrue("Closed Status", Connection.Status.CLOSED == consumer.getStatus());
    }

  }
```

**Comment by sasbury** [comment]
this makes sense, let me integrate the tests and think about how to get the auth through when the buffer is full. Thanks for digging on this!

**Comment by sasbury** [comment]
just looking at code the connect protocol message is getting stuck behind the buffer, so I will either put in your test, or just add auth to my buffer test or both and then fix the issue.

**Comment by Imaskar** [comment]
Thanks. Unfortunately,  my test is not deterministic. It hopes the comnsumer would subscribe before all 6k from producer's buffer will be sent. And I don't see how to postpone reconnecting of the producer.

**Comment by sasbury** [comment]
i did a test with many fewer messages, but with the timing and auth to make sure that the connect went through before the messages. Created a fix, and running tests now. If it works i will get it on github for review asap.

**Comment by Imaskar** [comment]
Your test like mine hopes that subscription will be reapplied before the buffer will be flushed. Before changes it was not always the case. I saw in the log at least one ping, sometimes more, so >1ms. Is it certain after the changes?

**Comment by Imaskar** [comment]
I checked if I could hold publisher from reconnecting by waiting on a semaphore in event listener, but it didn't work - reconnecting happens in another thread.

**Comment by sasbury** [comment]
my fix provides a fast path for all the protocol messages until the reconnect process is complete. So all the subscriptions are set up before the protocol messages start using the same queue as the regular messages. This is definitely something that would be very hard to work around, but I think the changes in v2.1.1 (branch is there if you want to look), should fix the issue because they allow the connect, ping and subscribe to jump the queue.
